### PR TITLE
add pids stats

### PR DESCRIPTION
**Pull request by KielChan**
We will collect pids metrics in kubelet by enable subcontainers to detect Burstable, Guaranteed and BestEffort pod pids stats.
And also it will be used to detect threads leaked if used jvm or some other virtual machine.
 The output maybe like this:
```
			"processes": {
				"process_count": 1,
				"fd_count": 8,
                                 "threads_current": 66,
                                 "threads_max": 0
			}
```

in `kubepods` and if subcontainers enabled.

`threads_max` will be detected in pod metrics `Processes` Spec `limit` field if pids_limit feature enabled in kubelet.

Ref: https://github.com/google/cadvisor/pull/1891

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
this is great to see.

fyi @sjenning @mrunalp @RobertKrawitz

File: metrics/prometheus.go [hunk 1]
```diff
@@ -415,6 +415,32 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 			},
 		}...)
 	}
+	if includedMetrics.Has(container.PidsMetrics) {
```
Comment (dashpole) [inline metrics/prometheus.go]: We already have container_processes.  How does this differ from that metric?

File: info/v1/container.go [hunk 2]
```diff
@@ -41,6 +41,10 @@ type MemorySpec struct {
 	SwapLimit uint64 `json:"swap_limit,omitempty"`
 }
 
+type PidsSpec struct {
+	Limit uint64 `json:"limit,omitempty"`
```
[the remark this answers is not in the document]
Comment (KielChan) [reply info/v1/container.go]: just ref 
```
type MemorySpec struct {
	// The amount of memory requested. Default is unlimited (-1).
	// Units: bytes.
	Limit uint64 `json:"limit,omitempty"`

	// The amount of guaranteed memory.  Default is 0.
	// Units: bytes.
	Reservation uint64 `json:"reservation,omitempty"`

	// The amount of swap space requested. Default is unlimited (-1).
	// Units: bytes.
	SwapLimit uint64 `json:"swap_limit,omitempty"`
}
```

its value equals to pids.max, and it is a static variable if used pids limit feature in kubelet

File: metrics/prometheus.go (hunk 1 above)
Comment (KielChan) [reply metrics/prometheus.go]: `ProcessCount`  just collect processes num, and it is not equals to threads num in container.

File: info/v1/container.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (KielChan) [reply info/v1/container.go]: just ref 
```
type MemorySpec struct {
	// The amount of memory requested. Default is unlimited (-1).
	// Units: bytes.
	Limit uint64 `json:"limit,omitempty"`

	// The amount of guaranteed memory.  Default is 0.
	// Units: bytes.
	Reservation uint64 `json:"reservation,omitempty"`

	// The amount of swap space requested. Default is unlimited (-1).
	// Units: bytes.
	SwapLimit uint64 `json:"swap_limit,omitempty"`
}
```

its value equals to pids.max, and it is a static variable if used pids limit feature in kubelet

[the remark this answers is not in the document]
Comment (KielChan) [reply info/v1/container.go]: just ref 
```
type MemorySpec struct {
	// The amount of memory requested. Default is unlimited (-1).
	// Units: bytes.
	Limit uint64 `json:"limit,omitempty"`

	// The amount of guaranteed memory.  Default is 0.
	// Units: bytes.
	Reservation uint64 `json:"reservation,omitempty"`

	// The amount of swap space requested. Default is unlimited (-1).
	// Units: bytes.
	SwapLimit uint64 `json:"swap_limit,omitempty"`
}
```

its value equals to pids.max, and it is a static variable if used pids limit feature in kubelet

File: metrics/prometheus.go (hunk 1 above)
Comment (KielChan) [reply metrics/prometheus.go]: `ProcessCount`  just collect processes num, and it is not equals to threads num in container.

File: metrics/prometheus.go [hunk 3]
```diff
@@ -415,6 +415,32 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
+		c.containerMetrics = append(c.containerMetrics, []containerMetric{
+			{
+				name:      "container_pids_max",
+				help:      "The number of container max pids, infinity if value is zero",
```
Comment (sjenning) [inline metrics/prometheus.go]: reword:
The maximum number of processes allowed in the container, infinity if value is zero

File: metrics/prometheus.go [hunk 4]
```diff
@@ -415,6 +415,32 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
+				getValues: func(s *info.ContainerStats) metricValues {
+					return metricValues{{
+						value:     float64(s.Pids.Max),
+						timestamp: s.Timestamp,
+					}}
+				},
+			},
+			{
+				name:      "container_pids_current",
+				help:      "The number of container current pids",
```
Comment (sjenning) [inline metrics/prometheus.go]: reword:
The current number of processes in the container

File: container/common/helpers.go [hunk 5]
```diff
@@ -143,7 +152,7 @@ func readString(dirpath string, file string) string {
 
 func readUInt64(dirpath string, file string) uint64 {
 	out := readString(dirpath, file)
-	if out == "" {
+	if out == "" || out == "max" {
```
Comment (sjenning) [inline container/common/helpers.go]: why is this needed?

Comment (sjenning) [reply container/common/helpers.go]: ah nm, `max` is the value in `/sys/fs/cgroup/pids/.../pids.max` when the max is unbounded :+1:

File: metrics/prometheus.go (hunk 1 above)
Comment (sjenning) [reply metrics/prometheus.go]: I think I would expect these to match.  `ProcessCount` counts the number of pids listed in the `cgroups.proc` file in the `cpu` hierarchy where the new stat just reads the `pids.current` in the `pids` hierarchy.

File: info/v1/container.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjenning) [reply info/v1/container.go]: This was also confusing to be at first, but I can see the rationale for the `Max` vs `Limit` as kubelet refers to this as `Limit` for all other resources. :+1:

**Comment by sjenning** [comment]
LGTM, other than some rewording nits in the in Prom metric descriptions

I do yield that we already have a metric for the number of processes in a container, but I think reading straight out of the pids cgroup, which is used for pid limit enforcement, is the better way.

File: metrics/prometheus.go (hunk 1 above)
Comment (KielChan) [reply metrics/prometheus.go]: pids.current equals to processes + threads num，its value is like pstree-l entries num

File: container/common/helpers.go (hunk 5 above)
Comment (KielChan) [reply container/common/helpers.go]: That's right.

File: metrics/prometheus.go (hunk 3 above)
Comment (KielChan) [reply metrics/prometheus.go]: I will fix it

Comment (KielChan) [reply metrics/prometheus.go]: I have fixed it

File: metrics/prometheus.go (hunk 4 above)
Comment (KielChan) [reply metrics/prometheus.go]: I have fixed it too

File: container/factory.go [hunk 6]
```diff
@@ -53,6 +53,7 @@ const (
 	NetworkUsageMetrics     MetricKind = "network"
 	NetworkTcpUsageMetrics  MetricKind = "tcp"
 	NetworkUdpUsageMetrics  MetricKind = "udp"
+	PidsMetrics             MetricKind = "pids"
```
Comment (dashpole) [inline container/factory.go]: Lets stick with ProcessMetrics

File: info/v1/container.go [hunk 7]
```diff
@@ -565,13 +572,19 @@ type ProcessStats struct {
...
+}
+
 type ContainerStats struct {
 	// The time of this stat point.
 	Timestamp time.Time    `json:"timestamp"`
 	Cpu       CpuStats     `json:"cpu,omitempty"`
 	DiskIo    DiskIoStats  `json:"diskio,omitempty"`
 	Memory    MemoryStats  `json:"memory,omitempty"`
 	Network   NetworkStats `json:"network,omitempty"`
+	Pids      PidsStats    `json:"pids,omitempty"`
```
Comment (dashpole) [inline info/v1/container.go]: Lets stick with ProcessStats

File: info/v1/container.go [hunk 8]
```diff
@@ -41,6 +41,10 @@ type MemorySpec struct {
 	SwapLimit uint64 `json:"swap_limit,omitempty"`
 }
 
+type PidsSpec struct {
```
Comment (dashpole) [inline info/v1/container.go]: Can we call this ProcessSpec for consistency?

File: info/v1/container.go [hunk 9]
```diff
@@ -565,13 +572,19 @@ type ProcessStats struct {
 	FdCount uint64 `json:"fd_count"`
 }
 
+type PidsStats struct {
```
Comment (dashpole) [inline info/v1/container.go]: Lets add these to ProcessStats

File: metrics/prometheus.go (hunk 1 above)
Comment (dashpole) [reply metrics/prometheus.go]: This is different from threads, right?  Or if this is a threads metric, lets name it that way

File: info/v1/container.go (hunk 8 above)
Comment (KielChan) [reply info/v1/container.go]: ok, I will move it

File: info/v1/container.go (hunk 9 above)
Comment (KielChan) [reply info/v1/container.go]: fine, I will fix it today

File: metrics/prometheus.go (hunk 1 above)
Comment (KielChan) [reply metrics/prometheus.go]: @dashpole it's threads metrics, I will rename it by prefix like `Threads(Max|Current)`

File: metrics/prometheus.go [hunk 10]
```diff
@@ -1045,7 +1045,30 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 					return metricValues{{value: float64(s.Processes.FdCount), timestamp: s.Timestamp}}
 				},
 			},
+			{
+				name:      "container_pids_max",
```
Comment (dashpole) [inline metrics/prometheus.go]: `container_threads_max`

File: metrics/prometheus.go [hunk 11]
```diff
@@ -1045,7 +1045,30 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
+				valueType: prometheus.GaugeValue,
+				getValues: func(s *info.ContainerStats) metricValues {
+					return metricValues{{
+						value:     float64(s.Processes.ThreadsMax),
+						timestamp: s.Timestamp,
+					}}
+				},
+			},
+			{
+				name:      "container_pids_current",
```
Comment (dashpole) [inline metrics/prometheus.go]: `container_threads`.  To be consistent with other metrics, omit the `_current` suffix.

File: metrics/prometheus.go (hunk 10 above)
Comment (KielChan) [reply metrics/prometheus.go]: fine, I will fix it.

File: container/raw/handler.go [hunk 12]
```diff
@@ -77,6 +77,9 @@ func newRawContainerHandler(name string, cgroupSubsystems *libcontainer.CgroupSu
 	pid := 0
 	if isRootCgroup(name) {
 		pid = 1
+
+		// delete pids from cgroup paths because /sys/fs/cgroup/pids/pids.current not exist
+		delete(cgroupPaths, "pids")
```
Comment (dashpole) [inline container/raw/handler.go]: This means we won't have current or max threads for the root cgroup, right?

That would be useful to have if we want to have the kubelet perform eviction when we are running low.  I don't think it needs to be a requirement for this PR, but we should consider adding this as a future improvement.  We would want to get the limit from `/proc/sys/kernel/threads-max`, and (probably?) would have to scan `/proc` for total thread count.
